src/main/java/fi/aalto/cs/apluscourses/intellij/actions/ImportProjectSettingsAction.java

```diff
@@ -0,0 +1,133 @@
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.util.io.FileUtilRt;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.UnexpectedResponseException;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+import fi.aalto.cs.apluscourses.ui.IntelliJDialogs;
+import fi.aalto.cs.apluscourses.ui.base.Dialogs;
+import fi.aalto.cs.apluscourses.utils.CoursesClient;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+import net.lingala.zip4j.ZipFile;
+import net.lingala.zip4j.model.FileHeader;
+import org.jetbrains.annotations.NotNull;
+
+public class ImportProjectSettingsAction extends AnAction {
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull ProjectSettingsImporter projectSettingsImporter;
+
+  @NotNull
+  private Dialogs dialogs;
+
+  /**
+   * Construct an action with the given main view model provider, project settings importer and
+   * dialog helper.
+   */
+  public ImportProjectSettingsAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                                     @NotNull ProjectSettingsImporter projectSettingsImporter,
+                                     @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.projectSettingsImporter = projectSettingsImporter;
+    this.dialogs = dialogs;
+  }
+
+  public ImportProjectSettingsAction() {
+    this(PluginSettings.getInstance(), ImportProjectSettingsAction::doImport,
+        new IntelliJDialogs());
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    Project project = e.getProject();
+    if (project == null) {
+      informNoProjectOpen();
+      return;
+    }
+
+    CourseViewModel courseViewModel = mainViewModelProvider
+        .getMainViewModel(project)
+        .courseViewModel
+        .get();
+    if (courseViewModel == null) {
+      return;
+    }
+    Course course = courseViewModel.getModel();
+
+    URL settingsUrl = course.getResourceUrls().get("projectSettings");
+    if (settingsUrl == null) {
+      informNoSettings(course);
+      return;
+    }
+
+    try {
+      projectSettingsImporter.doImport(project, settingsUrl);
+    } catch (IOException | UnexpectedResponseException ex) {
+      informErrorOccurred();
+    }
+  }
+
+  /**
+   * Downloads the project settings ZIP file from the given URL to a temporary file. After that the
+   * files from the .idea directory of the ZIP file are extracted to the .idea directory of the
+   * given project, after which the project is reloaded.
+   */
+  private static void doImport(@NotNull Project project, @NotNull URL settingsUrl)
```

Comment:
I wonder if this could be split into smaller methods?

Reply:
Good idea :+1: